Feature/custom account details

emclaug2 opened the pull request:
<!-- Include a bulleted list summarizing the main changes you have made in this PR -->
#### Changes proposed in this Pull Request:
- Removes ability to skip account details.
- Makes First/Last name mandatory.
- Changes the way custom account details are passed into the self-registration screens.
- Adds CreateAccountService (internal) to help with rendering logic. 
- TODO: update docs

This is how to pass in custom account details.  We need these three things:

#### 1. Custom `<pxb-create-account>` component.

Each time a page customization happens through `ng-content` or `@Inputs`, this step is required.   
```
 <!-- Custom Create Account page -->
<ng-template #createAccountPage>
    <pxb-create-account [accountDetails]="accountDetails"></pxb-create-account>
</ng-template>
```


#### 2.  Custom forms accessible as a `TemplateRef`. 

This is what to render as your custom form.
```
<ng-template #customFormRef>
  <form>
      <mat-form-field appearance="fill" [style.width.%]="100" [style.marginBottom.px]="8">
          <mat-label>Custom Field</mat-label>
          <input matInput [formControl]="customFormControl" required />
          <mat-error *ngIf="customFormControl.hasError('required')">
              Emergency Contact is <strong>required</strong>
          </mat-error>
      </mat-form-field>
  </form>
</ng-template>
```

#### 3.  An `accountDetails[]` `@Input`
Each `accountDetails` object has 3 properties:

`form`: what to render
`formControls`: access what the user entered
`isValid()`: function we run to determine if user-input is valid.

Each `accountDetails` represents a new page in the self-registration process.  
```
import { AccountDetails } from '@pxblue/angular-auth-workflow';


// What to Render
@ViewChild('customFormRef') customFormRef: TemplateRef<MatFormField>;

// To capture user inputs
customFormControl: FormControl;

// An array of accountDetails where each object represents a new page. 
accountDetails: AccountDetails[]; 

ngAfterViewInit(): void {
  this.customFormControl= new FormControl('', Validators.required);
  this.accountDetails = [
      {   /* Page 1 */
          form: this.customFormRef,
          formControls: [this.customFormControl],
          isValid: () => this.customFormControl.value,
      }
  ];
}

```

This solution scales so that a user can have as many pages as they want.  

Users can even have blank custom pages as well.  This is useful if they do want want any custom forms on the first account details page (first/last name).  e.g.

```
/* Page 1 */
{
    form: undefined, formControls: undefined, isValid: () => true
},
/* Page 2 */
{
    form: this.customFormRef,
    formControls: [this.customFormControl],
    isValid: () => this.customFormControl.value,
}

```

daileytj reviewed (approved):
Looks good! 

One minor thing though, when we're using custom form fields we lose the ability to progress and submit via "enter" keypress. If there's a simple way to handle this I would recommend adding that before merging with dev. @joebochill @huayunh thoughts?

emclaug2 commented:
I agree with you Tom that it would be cool to retain the next-on-enter functionality.  Let me see if I can figure out an easy way to do this.

emclaug2 commented:
Users can manually add the next-on-enter functionality.  If we build this into the example then it doesn't seem too inconvienient. 
```
<ng-template #createAccountPage>
  <pxb-create-account #createAccount [accountDetails]="accountDetails"></pxb-create-account>
</ng-template>
```
```
<mat-form-field appearance="fill" [style.width.%]="100" [style.marginBottom.px]="8">
  <mat-label>Phone Number (optional)</mat-label>
  <input matInput [formControl]="phoneNumberFormControl"
         (keyup.enter)="createAccount.canContinue() ? createAccount.goNext() : ''"/>
</mat-form-field>
```
```
@ViewChild('createAccount') createAccount: PxbCreateAccountComponent;
```

These ViewChild names are not very intuitive so I'm going to tweak them a bit to make more sense.  Also this logic can be placed into a single function:

```
 (keyup.enter)="createAccount.canContinue() ? createAccount.goNext() : ''"
```

joebochill reviewed (commented):
I think this needs some documentation in the readme around how to do the customization. Also can you give a demo of this sometime today or tomorrow so I can see how it works?

emclaug2 commented:
> I think this needs some documentation in the readme around how to do the customization. Also can you give a demo of this sometime today or tomorrow so I can see how it works?

Yeah we can do a demo either today or tomorrow in office hours.  It's at a demo-able point.

emclaug2 commented:
Couple updates:
- Added docs on how to add custom form fields to self registration
- Make those [style.margin.px] and [style.width.%] styles **not** required 
- Added next-on-enter functionality to example project custom forms
- Added country code to custom form field. 
- Changed data structure of `accountDetails` to more-easily access user-provided data. 

![image](https://user-images.githubusercontent.com/6538289/106801794-2b8c2400-6630-11eb-9a47-bd6c1dcfe20c.png)

joebochill commented on login-workflow/README.md:
````diff
@@ -400,6 +400,74 @@ See the example project (`./src/app/dialog/login-error-dialog.component.ts`) for
+
+```
+
+This solution scales so that a user can have as many pages as they want.  Adding `undefined` as the first entry of the `accountDetails[]` will shift all custom forms onto their own page. 
````
Maybe add a shout out to the demo project "Check out the <link to> Example project to see an example of custom registration forms"
